spring-kafka/src/main/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainer.java

```diff
@@ -235,7 +243,7 @@ public boolean isChildRunning() {
 	 */
 	@Override
 	protected void doStart() {
-		if (!isRunning()) {
+		if (!isRunning() && this.containers.stream().allMatch(container -> !container.isRunning())) {
```

Comment:
This again sounds like a contradiction to what we have with fenced child container.
Why would one be running if parent is stopped?

Reply:
This condition is to prevent `run` call on ConcurrentContainer before stopping the child containers. This is to verify if really all the containers are stopped.

This is similar to earlier logic. If container is in running status , subsequent `run` calls will be ignored.

Earlier, running status of the concurrent container is not set to false, even if one of the container is stopped. I have made the changes to set  concurrent container running status to false, even if one of the container is stopped. 
   
  In this case, condition to verify if concurrent is allowed to start needs to be changed. If all the containers running status is set to 'false' then only `run` is allowed.

Reply:
> In this case, condition to verify if concurrent is allowed to start needs to be changed. If all the containers running status is set to 'false' then only run is allowed.

That's also not what I think about this concurrent container logic.
It is in running state when any of its child containers running.
When we start concurrent container, it suppose to start all of its children.
If some of them are running already, then idempotent.
As we state before: we just don't allow orphaned child containers to be restarted.
Probably much robust solution is to stop all the children when we call start of their parent.
This way any new start would give us a fresh state. Kinda total renew.

Reply:
> When we start concurrent container, it suppose to start all of its children.
> If some of them are running already, then idempotent.
> As we state before: we just don't allow orphaned child containers to be restarted.
> Probably much robust solution is to stop all the children when we call start of their parent.
> This way any new start would give us a fresh state. Kinda total renew.

This is already happening. I have not modified any thing here.

Reply:
> That's also not what I think about this concurrent container logic.
> It is in running state when any of its child containers running.

I have replied to this comment. I have set the `running` status to `false` as per the API definition

Reply:
I see that you refer to `Lifecycle.isRunning()`.
I'm not sure what that `container` means there, but that does not compile in my head.
How we suppose to treat container as not running if any of its child still does its work?
That really means we have a leak.

I still believe that we have to return `true` if any of the children are running.
Or leave it as is, because we just don't change concurrent `running` state even if we stop all its children manually.

Reply:
> I still believe that we have to return `true` if any of the children are running.

I am good with this. 
But one query, what is your suggestion to know from the API, if any of the container is stopped. Do we need to get all the containers and verify if any of it is stopped.

Reply:
> Or leave it as is, because we just don't change concurrent `running` state even if we stop all its children manually.

I think here we must set the `running` status to `false` if all the containers are stopped. Let us not leave as it is.

Reply:
Well, that's `start()` of the concurrent container.
I think the best solution would be to stop all the currently running and go ahead with the rest of the logic where we really re-create child container. 

I might agree that we can move concurrent container to not running state if all of its children are stopped.
If we have an API to notify via `thisOrParentContainer` that child is stopped.

> Do we need to get all the containers and verify if any of it is stopped.

Why do we need to check if they are stopped?
I think iteration over `this.containers` and checking `isRunning()` is enough.
Is `isChildRunning()` OK?

Reply:
> Well, that's `start()` of the concurrent container.
> I think the best solution would be to stop all the currently running and go ahead with the rest of the logic where we really re-create child container.

As per the current code,

Start API would start new child containers.
Stop API would stop any containers running.

I feel it is correct. Let us not change any thing regarding this. I have added new conditions verify if really ConcurrentContainer is really stopped or not.

As per the earlier code, second `run` is ignored. APP dev can directly `stop` and `run` the ConcurrentContainer if he plans to run again. But, good practise to wait untill `isChildRunningAPI` returns `false`. This indicates all the resources are released.